src/platform/datapath_win.c

```diff
@@ -0,0 +1,636 @@
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return NULL;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSendDataFreeBuffer(
+    _In_ CXPLAT_SEND_DATA* SendData,
+    _In_ QUIC_BUFFER* Buffer
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        SendDataFreeBuffer(SendData, Buffer);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        RawSendDataFreeBuffer(SendData, Buffer);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+CxPlatSendDataIsFull(
+    _In_ CXPLAT_SEND_DATA* SendData
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        return SendDataIsFull(SendData);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        return RawSendDataIsFull(SendData);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return FALSE;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+CxPlatSocketSend(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_ const CXPLAT_ROUTE* Route,
+    _In_ CXPLAT_SEND_DATA* SendData
+    )
+{
+    if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_USER) {
+        return SocketSend(Socket, Route, SendData);
+    } else if (DatapathType(SendData) == CXPLAT_DATAPATH_TYPE_XDP) {
+        return RawSocketSend(SocketToRaw(Socket), Route, SendData);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+    return QUIC_STATUS_NOT_SUPPORTED;
+}
+
+void
+CxPlatDataPathProcessCqe(
+    _In_ CXPLAT_CQE* Cqe
+    )
+{
+    // TODO: refactoring
+    switch (CxPlatCqeType(Cqe)) {
+    case CXPLAT_CQE_TYPE_SOCKET_IO: {
+        DATAPATH_IO_SQE* Sqe =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), DATAPATH_IO_SQE, DatapathSqe);
+        if (Sqe->IoType == DATAPATH_XDP_IO_RECV || Sqe->IoType == DATAPATH_XDP_IO_SEND) {
+            RawDataPathProcessCqe(Cqe);
+        } else {
+            DataPathProcessCqe(Cqe);
+        }
+        break;
+    }
+    case CXPLAT_CQE_TYPE_SOCKET_SHUTDOWN: {
+        RawDataPathProcessCqe(Cqe);
+        break;
+    }
+    default: CXPLAT_DBG_ASSERT(FALSE); break;
+    }
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicCopyRouteInfo(
+    _Inout_ CXPLAT_ROUTE* DstRoute,
+    _In_ CXPLAT_ROUTE* SrcRoute
+    )
+{
+    if (SrcRoute->DatapathType == CXPLAT_DATAPATH_TYPE_XDP) {
+        CxPlatCopyMemory(DstRoute, SrcRoute, (uint8_t*)&SrcRoute->State - (uint8_t*)SrcRoute);
+        CxPlatUpdateRoute(DstRoute, SrcRoute);
+    } else if (SrcRoute->DatapathType == CXPLAT_DATAPATH_TYPE_USER) {
+        *DstRoute = *SrcRoute;
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+}
+
+void
+CxPlatResolveRouteComplete(
+    _In_ void* Connection,
+    _Inout_ CXPLAT_ROUTE* Route,
+    _In_reads_bytes_(6) const uint8_t* PhysicalAddress,
+    _In_ uint8_t PathId
+    )
+{
+    if (Route->State != RouteResolved) {
+        RawResolveRouteComplete(Connection, Route, PhysicalAddress, PathId);
```

Comment:
Debug assert that the type is RAW/XDP?

Reply:
Allowing to call from any datapath, but functionality is basically for RAW